compiler/frontend/src/org/jetbrains/kotlin/implicit/ImplicitResolutionStrategy.kt

```
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.implicit
+
+import org.jetbrains.kotlin.implicit.ImplicitCandidateResolution.Resolved
+import org.jetbrains.kotlin.implicit.ImplicitCandidateResolution.Unresolved
+import org.jetbrains.kotlin.implicit.ImplicitResolution.*
+import org.jetbrains.kotlin.descriptors.ValueParameterDescriptor
+import org.jetbrains.kotlin.resolve.calls.model.ImplicitValueArgument
+
+object ImplicitResolutionStrategy {
+    @JvmStatic
+    fun resolve(
+        lookingFor: ValueParameterDescriptor,
+        parameters: List<ValueParameterDescriptor>,
+        argumentParameterDescriptor: ValueParameterDescriptor,
+        substitutions: List<TypeSubstitution>,
+        lookInSupertypes: Boolean = false
+    ): ImplicitCandidateResolution {
+
+        val functionOrder = listOf(
+            FindInLocalFunction,
+            FindInPackage,
+            FindInTypeCompanion,
+            FindInTypeclassCompanion,
+            FindInTypeSubpackages,
+            FindInTypeclassSubpackages
+        )
+        val candidate = functionOrder.fold(Unresolved("") as ImplicitCandidateResolution) { acc, resolution ->
+            when (acc) {
+                !is Resolved -> resolution.resolve(lookingFor, parameters, argumentParameterDescriptor, substitutions, lookInSupertypes)
+                else -> acc
```

Fail fast (as soon as there's one resolved it's carried til the end, we want a single possible resolution).